control-plane/rest/src/lib.rs

```diff
@@ -148,51 +161,160 @@ impl ActixRestClient {
-    ) -> anyhow::Result<R>
+    ) -> Result<R, ClientError>
     where
         S: Stream<Item = Result<Bytes, PayloadError>> + Unpin,
         for<'de> R: Deserialize<'de>,
     {
-        let rest_body = rest_response.body().await?;
-        if rest_response.status().is_success() {
-            Ok(serde_json::from_slice::<R>(&rest_body)?)
+        let status = rest_response.status();
+        let headers = rest_response.headers().clone();
+        let head = || {
+            let mut head = ResponseHead::new(status);
+            head.headers = headers.clone();
+            head
+        };
+        let body = rest_response.body().await.map_err(|_| {
+            ClientError::InvalidPayload {
+                head: head(),
+            }
+        })?;
+        if status.is_success() {
+            let result = serde_json::from_slice(&body).map_err(|_| {
+                ClientError::InvalidBody {
+                    head: head(),
+                    body,
+                }
+            })?;
+            Ok(result)
+        } else if body.is_empty() {
+            Err(ClientError::Header {
+                head: head(),
+            })
         } else {
-            let error: serde_json::value::Value =
-                serde_json::from_slice(&rest_body)?;
-            Err(anyhow::anyhow!(error.to_string()))
+            let error = serde_json::from_slice::<serde_json::Value>(&body)
+                .map_err(|_| ClientError::InvalidBody {
+                    head: head(),
+                    body,
+                })?;
+            Err(ClientError::Valid {
+                head: head(),
+                error,
+            })
+        }
+    }
+}
+
+/// Result of a Rest Client Operation
+/// T is the Object parsed from the Json body
+pub type ClientResult<T> = Result<T, ClientError>;
+
+/// Rest Client Error
+#[derive(Debug, Snafu)]
+pub enum ClientError {
+    /// Failed to send message to the server
+    #[snafu(display("{}, reason: {}", details, source))]
+    Send {
+        /// Message
+        details: String,
+        /// Source Request Error
+        source: SendRequestError,
+    },
+    /// Invalid Resource Filter
+    #[snafu(display("Invalid Resource Filter: {}", details))]
+    InvalidFilter {
+        /// Message
+        details: String,
+    },
+    /// Invalid Payload
+    #[snafu(display("Invalid payload, header: {:?}", head))]
+    InvalidPayload {
+        /// http Header
+        head: ResponseHead,
+    },
+    /// Invalid Body
+    #[snafu(display("Invalid body, header: {:?}", head))]
+    InvalidBody {
+        /// http Header
+        head: ResponseHead,
+        /// http Body
+        body: Bytes,
+    },
+    /// No Body
+    #[snafu(display("No body, header: {:?}", head))]
+    Header {
+        /// http Header
+        head: ResponseHead,
+    },
+    /// Body in JSON format
+    #[snafu(display("Http status: {}, error: {}", head.status, error.to_string()))]
+    Valid {
```

Comment:
Could we rename `Valid` to something else? It seems a little strange to have a valid error.

Reply:
Yeah, it was valid in the sense that it's a valid json format error. Any suggestions?

Reply:
Renamed to `RestServer`

Reply:
:+1: